sage.py

```diff
@@ -481,1274 +423,920 @@ def getepisodes(action_seq, mcat, plot, debug=False):
+        state = dfa[state][sym]
+        if state == "":
+            # Only keep IDs of medium- and high-severity sinks
+            if len(str(sev)) >= 2:
+                # TODO: a better way is `if state_list[-1] in sinks and sym in sinks[state_list[-1]]:`
+                # However, in this case `-1` will not be created in `sinks` when accessing a non-existent key
+                # This is an issue when state_list[-1] == '-1' (e.g. in CCDC-2018)
+                # Alternatively, change defaultdict to dict
                 try:
-                    state = sinks[stlst[-1]][sym][0]
-                except:
-                    #print('didnt work for', sequence, 'done so far:', stlst)
-                    state = '-1'
-          else:
-              state = '-1'
-          #print("Must look in sinks")
-          #print('prev symb: ', sym, 'and prev state no.', stlst[-1])
-          #print('BUT this last state no in sink gives ', sinks[stlst[-1]][sym])
-          #print('curr sym', sym)
-          #print('previous state no.:', stlst[-1], 'results in sink id:', sinks[stlst[-1]][sym] )
-          #if sinks[stlst[-1]][sym] == "":
-                
-                #print('prob')
-                #print(sequence)
-                #state = '-1'
-          #else:
-          #      state = sinks[stlst[-1]][sym][0]
-          #
-          #if not statelist:
-          #        return dfa[state]["type"] == "1"
-          #else:
-          #        return (dfa[state]["type"] == "1", stlst)
-
-      else:
-          try:
-              #print('weird place')
-              # take target id, discard counts
-              state = state[0]
-          except IndexError:
-              print("Out of alphabet: alternatives")
-          
-              stlst.append("-1")
-              if not statelist:
-                     return dfa[state]["type"] == "1"
-              else:
-                     return (dfa[state]["type"] == "1", stlst)
-      stlst.append(state)
-      if state in sinks and len(str(sev)) >= 2:
-          sev_sinks.add(state)
-  if not statelist:
-      return dfa[state]["type"] == "1"
-  else:
-      return (dfa[state]["type"] == "1", stlst, sev_sinks)
-
-def encode_sequences(m, m2):
-    #print(m2)
-    traces = []
-    sp = []
-    orig = []
-    with open(path_to_traces) as tf:
-        lines = tf.readlines()[1:]
-    #print(len(lines))
-
-    for line in lines:
-        if line == lines[-1]:
-             spl = line.split(' ')
-        else:
-            spl = line[:-1].split(' ')
-        
-        line = ' '.join(spl[2:])
-        #print(spl[2:])
-        orig.append([(x) for x in spl[2:] if x != ''])
-        traces.append(line)
-    num_sink = 0   
-    total = 0
+                    state = sinks[state_list[-1]][sym]
+                    state = '-1' if state == '' else state
+                except IndexError:
+                    state = '-1'  # With `printblue = 1` in spdfa-config.ini this should not happen
```

Comment:
...and after what I have written in the comment above, though this part can still be kept